File: Catalog core/src/test/java/cz/vhromada/catalog/repository/MusicRepositoryIntegrationTest.java
package cz.vhromada.catalog.repository;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.commons.MusicUtils;
import cz.vhromada.catalog.commons.SongUtils;
import cz.vhromada.catalog.dao.entities.Music;
import cz.vhromada.catalog.dao.entities.Song;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.Sort;
import org.springframework.test.annotation.Rollback;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

/**
 * A class represents integration test for class {@link MusicRepository}.
 *
 * @author Vladimir Hromada
 */
@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration("classpath:testRepositoryContext.xml")
@Transactional
@Rollback
public class MusicRepositoryIntegrationTest {

    /**
     * Instance of {@link EntityManager}
     */
    @Autowired
    @Qualifier("containerManagedEntityManager")
    private EntityManager entityManager;

    /**
     * Instance of {@link MusicRepository}
     */
    @Autowired
    private MusicRepository musicRepository;

    /**
     * Test method for get all music.
     */
    @Test
    public void testGetMusic_All() {
        final List<Music> music = musicRepository.findAll(new Sort("position", "id"));

        MusicUtils.assertMusicDeepEquals(MusicUtils.getMusic(), music);

        assertEquals(MusicUtils.MUSIC_COUNT, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for get one music.
     */
    @Test
    public void testGetMusic_One() {
        for (int i = 1; i <= MusicUtils.MUSIC_COUNT; i++) {
            final Music music = musicRepository.findOne(i);

            MusicUtils.assertMusicDeepEquals(MusicUtils.getMusic(i), music);
        }

        assertNull(musicRepository.findOne(Integer.MAX_VALUE));

        assertEquals(MusicUtils.MUSIC_COUNT, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for add music.
     */
    @Test
    public void testAdd() {
        final Music music = MusicUtils.newMusic(null);

        musicRepository.saveAndFlush(music);

        assertNotNull(music.getId());
        assertEquals(MusicUtils.MUSIC_COUNT + 1, music.getId().intValue());

        final Music addedMusic = MusicUtils.getMusic(entityManager, MusicUtils.MUSIC_COUNT + 1);
        final Music expectedAddMusic = MusicUtils.newMusic(null);
        expectedAddMusic.setId(MusicUtils.MUSIC_COUNT + 1);
        MusicUtils.assertMusicDeepEquals(expectedAddMusic, addedMusic);

        assertEquals(MusicUtils.MUSIC_COUNT + 1, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for update music with updated data.
     */
    @Test
    public void testUpdate_Data() {
        final Music music = MusicUtils.updateMusic(entityManager, 1);

        musicRepository.saveAndFlush(music);

        final Music updatedMusic = MusicUtils.getMusic(entityManager, 1);
        final Music expectedUpdatedMusic = MusicUtils.getMusic(1);
        MusicUtils.updateMusic(expectedUpdatedMusic);
        expectedUpdatedMusic.setPosition(MusicUtils.POSITION);
        MusicUtils.assertMusicDeepEquals(expectedUpdatedMusic, updatedMusic);

        assertEquals(MusicUtils.MUSIC_COUNT, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for update music with added song.
     */
    @Test
    public void testUpdate_Song() {
        final Song song = SongUtils.newSong(null);
        entityManager.persist(song);

        final Music music = MusicUtils.getMusic(entityManager, 1);
        music.getSongs().add(song);

        musicRepository.saveAndFlush(music);

        final Music updatedMusic = MusicUtils.getMusic(entityManager, 1);
        final Song expectedSong = SongUtils.newSong(null);
        expectedSong.setId(SongUtils.SONGS_COUNT + 1);
        final Music expectedUpdatedMusic = MusicUtils.getMusic(1);
        expectedUpdatedMusic.getSongs().add(expectedSong);
        MusicUtils.assertMusicDeepEquals(expectedUpdatedMusic, updatedMusic);

        assertEquals(MusicUtils.MUSIC_COUNT, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT + 1, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for remove music.
     */
    @Test
    public void testRemove() {
        final int songsCount = MusicUtils.getMusic(1).getSongs().size();

        musicRepository.delete(MusicUtils.getMusic(entityManager, 1));

        assertNull(MusicUtils.getMusic(entityManager, 1));

        assertEquals(MusicUtils.MUSIC_COUNT - 1, MusicUtils.getMusicCount(entityManager));
        assertEquals(SongUtils.SONGS_COUNT - songsCount, SongUtils.getSongsCount(entityManager));
    }

    /**
     * Test method for remove all music.
     */
    @Test
    public void testRemoveAll() {
        musicRepository.deleteAll();

        assertEquals(0, MusicUtils.getMusicCount(entityManager));
        assertEquals(0, SongUtils.getSongsCount(entityManager));
    }

}

File: catalog-core/src/test/kotlin/com/github/vhromada/catalog/repository/MusicRepositoryIntegrationTest.kt
package com.github.vhromada.catalog.repository

import com.github.vhromada.catalog.CatalogTestConfiguration
import com.github.vhromada.catalog.utils.AuditUtils
import com.github.vhromada.catalog.utils.MusicUtils
import com.github.vhromada.catalog.utils.SongUtils
import com.github.vhromada.catalog.utils.TestConstants
import com.github.vhromada.catalog.utils.fillAudit
import com.github.vhromada.catalog.utils.updated
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.Rollback
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

/**
 * A class represents integration test for class [MusicRepository].
 *
 * @author Vladimir Hromada
 */
@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [CatalogTestConfiguration::class])
@Transactional
@Rollback
class MusicRepositoryIntegrationTest {

    /**
     * Instance of [EntityManager]
     */
    @PersistenceContext
    private lateinit var entityManager: EntityManager

    /**
     * Instance of [MusicRepository]
     */
    @Autowired
    private lateinit var repository: MusicRepository

    /**
     * Test method for get list of music.
     */
    @Test
    fun getMusicList() {
        val musicList = repository.findAll()

        MusicUtils.assertDomainMusicDeepEquals(expected = MusicUtils.getMusicList(), actual = musicList)

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

    /**
     * Test method for get music.
     */
    @Test
    fun getMusic() {
        for (i in 1..MusicUtils.MUSIC_COUNT) {
            val music = repository.findById(i).orElse(null)

            MusicUtils.assertMusicDeepEquals(expected = MusicUtils.getMusicDomain(index = i), actual = music)
        }

        assertThat(repository.findById(Int.MAX_VALUE)).isNotPresent

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

    /**
     * Test method for add music.
     */
    @Test
    fun add() {
        val music = MusicUtils.newMusicDomain(id = null)
            .copy(position = MusicUtils.MUSIC_COUNT)
        val expectedMusic = MusicUtils.newMusicDomain(id = MusicUtils.MUSIC_COUNT + 1)
            .fillAudit(audit = AuditUtils.newAudit())

        repository.save(music)

        assertSoftly {
            it.assertThat(music.id).isEqualTo(MusicUtils.MUSIC_COUNT + 1)
            it.assertThat(music.createdUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(music.createdTime).isEqualTo(TestConstants.TIME)
            it.assertThat(music.updatedUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(music.updatedTime).isEqualTo(TestConstants.TIME)
        }

        val addedMusic = MusicUtils.getMusic(entityManager, MusicUtils.MUSIC_COUNT + 1)!!
        assertThat(addedMusic).isNotNull
        MusicUtils.assertMusicDeepEquals(expected = expectedMusic, actual = addedMusic)

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT + 1)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

    /**
     * Test method for update music.
     */
    @Test
    fun update() {
        val music = MusicUtils.updateMusic(entityManager = entityManager, id = 1)
        val expectedMusic = MusicUtils.getMusicDomain(index = 1)
            .updated()
            .copy(position = MusicUtils.POSITION)
            .fillAudit(audit = AuditUtils.updatedAudit())

        repository.saveAndFlush(music)

        val updatedMusic = MusicUtils.getMusic(entityManager = entityManager, id = 1)
        assertThat(updatedMusic).isNotNull
        MusicUtils.assertMusicDeepEquals(expected = expectedMusic, actual = updatedMusic!!)

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

    /**
     * Test method for remove music.
     */
    @Test
    fun remove() {
        repository.delete(MusicUtils.getMusic(entityManager = entityManager, id = 1)!!)

        assertThat(MusicUtils.getMusic(entityManager = entityManager, id = 1)).isNull()

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT - 1)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT - SongUtils.SONGS_PER_MUSIC_COUNT)
        }
    }

    /**
     * Test method for remove all music.
     */
    @Test
    fun removeAll() {
        repository.deleteAll()

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(0)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(0)
        }
    }

    /**
     * Test method for get list of music for user.
     */
    @Test
    fun findByCreatedUser() {
        val musicList = repository.findByCreatedUser(user = AuditUtils.getAudit().createdUser!!)

        MusicUtils.assertDomainMusicDeepEquals(expected = MusicUtils.getMusicList(), actual = musicList)

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

    /**
     * Test method for get music by id for user.
     */
    @Test
    fun findByIdAndCreatedUser() {
        val user = AuditUtils.getAudit().createdUser!!
        for (i in 1..MusicUtils.MUSIC_COUNT) {
            val author = repository.findByIdAndCreatedUser(id = i, user = user).orElse(null)

            MusicUtils.assertMusicDeepEquals(expected = MusicUtils.getMusicDomain(index = i), actual = author)
        }

        assertThat(repository.findByIdAndCreatedUser(id = Int.MAX_VALUE, user = user)).isNotPresent

        assertSoftly {
            it.assertThat(MusicUtils.getMusicCount(entityManager = entityManager)).isEqualTo(MusicUtils.MUSIC_COUNT)
            it.assertThat(SongUtils.getSongsCount(entityManager = entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
        }
    }

}
